source/SAMRAI/geom/CartesianCellDoubleConservativeLinearRefine.C

```diff
@@ -82,247 +83,364 @@ void SAMRAI_F77_FUNC(cartclinrefcelldoub3d, CARTCLINREFCELLDOUB3D) (const int&,
-         coarse.getPatchData(src_component)));
+       SAMRAI_SHARED_PTR_CAST<pdat::CellData<double>, hier::PatchData>(
+           coarse.getPatchData(src_component)));
    std::shared_ptr<pdat::CellData<double> > fdata(
-      SAMRAI_SHARED_PTR_CAST<pdat::CellData<double>, hier::PatchData>(
-         fine.getPatchData(dst_component)));
+       SAMRAI_SHARED_PTR_CAST<pdat::CellData<double>, hier::PatchData>(
+           fine.getPatchData(dst_component)));
    TBOX_ASSERT(cdata);
    TBOX_ASSERT(fdata);
    TBOX_ASSERT(cdata->getDepth() == fdata->getDepth());
 
    const hier::Box cgbox(cdata->getGhostBox());
 
-   const hier::Index& cilo = cgbox.lower();
-   const hier::Index& cihi = cgbox.upper();
-   const hier::Index& filo = fdata->getGhostBox().lower();
-   const hier::Index& fihi = fdata->getGhostBox().upper();
+   const hier::Index &cilo = cgbox.lower();
+   const hier::Index &cihi = cgbox.upper();
+   const hier::Index &filo = fdata->getGhostBox().lower();
+   const hier::Index &fihi = fdata->getGhostBox().upper();
 
    const std::shared_ptr<CartesianPatchGeometry> cgeom(
-      SAMRAI_SHARED_PTR_CAST<CartesianPatchGeometry, hier::PatchGeometry>(
-         coarse.getPatchGeometry()));
+       SAMRAI_SHARED_PTR_CAST<CartesianPatchGeometry, hier::PatchGeometry>(
+           coarse.getPatchGeometry()));
    const std::shared_ptr<CartesianPatchGeometry> fgeom(
-      SAMRAI_SHARED_PTR_CAST<CartesianPatchGeometry, hier::PatchGeometry>(
-         fine.getPatchGeometry()));
+       SAMRAI_SHARED_PTR_CAST<CartesianPatchGeometry, hier::PatchGeometry>(
+           fine.getPatchGeometry()));
 
    TBOX_ASSERT(cgeom);
    TBOX_ASSERT(fgeom);
 
    const hier::Box coarse_box = hier::Box::coarsen(fine_box, ratio);
-   const hier::Index& ifirstc = coarse_box.lower();
-   const hier::Index& ilastc = coarse_box.upper();
-   const hier::Index& ifirstf = fine_box.lower();
-   const hier::Index& ilastf = fine_box.upper();
+   const hier::Index &ifirstc = coarse_box.lower();
+   const hier::Index &ilastc = coarse_box.upper();
+   const hier::Index &ifirstf = fine_box.lower();
+   const hier::Index &ilastf = fine_box.upper();
 
    const hier::IntVector tmp_ghosts(dim, 0);
 
-
    SAMRAI::hier::Box diff_box = coarse_box;
    diff_box.growUpper(SAMRAI::hier::IntVector::getOne(dim));
 
-   tbox::AllocatorDatabase* alloc_db = tbox::AllocatorDatabase::getDatabase();
+   tbox::AllocatorDatabase *alloc_db = tbox::AllocatorDatabase::getDatabase();
 
-   pdat::CellData<double> diff(diff_box, dim.getValue(), tmp_ghosts, alloc_db->getDevicePool());
-   pdat::CellData<double> slope(cgbox, dim.getValue(), tmp_ghosts, alloc_db->getDevicePool());
+   //pdat::CellData<double> diff(diff_box, dim.getValue(), tmp_ghosts, alloc_db->getDevicePool());
+   //pdat::CellData<double> slope(cgbox, dim.getValue(), tmp_ghosts, alloc_db->getDevicePool());
+   pdat::CellData<double> diff(diff_box, dim.getValue(), tmp_ghosts, alloc_db->getTagAllocator());
```

Comment:
I don't think we want these to be using the tag allocator. The devicePool should work better. We might want to add something like a "getTemporaryAllocator" that switches between host/device depending on how SAMRAI is configured.

Reply:
Sounds good! Tag allocator was a convenient host-side allocator until we have some generic constructs.